src/MicrosoftAuthFlow.js

```diff
@@ -0,0 +1,150 @@
+const fs = require('fs');
+const path = require('path')
+const crypto = require('crypto');
+
+const minecraftFolderPath = require('minecraft-folder-path');
+const debug = require('debug')('minecraft-protocol');
+
+const { Authentication, msalConfig } = require('./Constants');
+
+async function retry(methodFn, beforeRetry, times) {
+    while (times--) {
+        if (times !== 0) {
+            try { return await methodFn() } catch (e) { debug(e) }
+            await new Promise(resolve => setTimeout(resolve, 2000))
+            await beforeRetry()
+        } else {
+            return await methodFn()
+        }
+    }
+}
+
+class MicrosoftAuthFlow {
+    constructor(username, cacheDir, options = {}, codeCallback) {
+        this.options = options
+        this.initTokenCaches(username, cacheDir)
+        this.codeCallback = codeCallback
+    }
+
+    initTokenCaches(username, cacheDir = mcDefaultFolderPath) {
+        const hash = sha1(username).substr(0, 6)
+
+        const cache = path.join(cacheDir, 'npm-cache');
+        debug(`Using cache path: ${cache}`)
+        try {
+            if (!fs.existsSync(cache)) {
+                fs.mkdirSync(cache, { recursive: true })
+            }
+        } catch (e) {
+            console.log('Failed to open cache dir', e)
+            cachePath = __dirname
+        }
+
+        const cachePaths = {
+            live: path.join(cachePath, `./${hash}_live-cache.json`),
+            msa: path.join(cachePath, `./${hash}_msa-cache.json`),
+            xbl: path.join(cachePath, `./${hash}_xbl-cache.json`),
+            bed: path.join(cachePath, `./${hash}_bed-cache.json`)
+        }
+
+        if (this.options.authTitle) { // Login with login.live.com
+            const scopes = ['service::user.auth.xboxlive.com::MBI_SSL']
+            this.msa = new LiveTokenManager(this.options.authTitle, scopes, cachePaths.live)
+        } else { // Login with microsoftonline.com
+            const scopes = ['XboxLive.signin', 'offline_access']
+            this.msa = new MsaTokenManager(msalConfig, scopes, cachePaths.msa)
+        }
+
+        const keyPair = crypto.generateKeyPairSync('ec', { namedCurve: 'P-256' })
+        this.xbl = new XboxTokenManager(Authentication.XSTSRelyingParty, keyPair, cachePaths.xbl)
+        this.mca = new MinecraftTokenManager(cachePaths.bed)
```

Comment:
need a good way to support both editions, not sure if we want to initialize everything here, maybe only based on the passed options

Reply:
We could probably promisify the function or return a callback once were done authenticating. returns the Xbox profile?

Reply:
It's already async promise no?